### PR TITLE
schema.sql: Add more constraints

### PR DESCRIPTION
**Pull request by RyanSquared**
Added the following constraints:
- check_hosts_for_hosts(): Check whether hosts have too many users
  before changing the maximum user limit
- check_invalid_name_for_passwd(): Checks whether a username exists in
  a group before inserting into passwd
- check_invalid_name_for_group(): Checks whether a group name exists in
  a passwd entry before inserting into group
- check_banned_username(): Check whether a username has been banned,
  such as if it's a name reserved for sysadmin usernames

Fixes: Part of #17, #27, #28.

### CODE REVIEW

File: schema.sql [hunk 1]
```diff
@@ -62,6 +75,49 @@ create constraint trigger max_users
...
+	return new;
+    end $$;
+create constraint trigger check_name_exists_passwd
+    after insert or update on passwd
+    for each row
+    execute procedure check_invalid_name_for_passwd();
+create constraint trigger check_name_exists_group
+    after insert or update on "group"
+    for each row
+    execute procedure check_invalid_name_for_group();
```
Comment (daurnimator) [inline schema.sql:106]: Could still have `when (new.name <> old.name)`

File: schema.sql [hunk 2]
```diff
@@ -62,6 +75,49 @@ create constraint trigger max_users
...
+
+-- prevent users from taking names typically reserved for sysadmins
+CREATE TABLE "banned_usernames" (
```
Comment (daurnimator) [inline schema.sql]: different case to the rest.

Comment (RyanSquared) [reply schema.sql]: I followed the case used when creating other tables.

Comment (daurnimator) [reply schema.sql]: oh. we should fix them all at once then.

**Comment by daurnimator** [comment]
>```
> ERROR:  INSERT trigger's WHEN condition cannot reference OLD values at character 116
> STATEMENT:  create constraint trigger check_name_exists_passwd
> 	    after insert or update on passwd
> 	    for each row
> 	    when (old.name <> new.name)
> 	    execute procedure check_invalid_name_for_passwd();
> psql:schema.sql:99: ERROR:  INSERT trigger's WHEN condition cannot reference OLD values
> LINE 4:     when (old.name <> new.name)
>                   ^
> ```

Should do: `tg_op = 'INSERT' or old.name <> new.name`

**Comment by RyanSquared** [comment]
`tg_op` doesn't exist at the time of `when`, logic will have to go in the function

**Comment by RyanSquared** [comment]
Reverted commit due to `tg_op` not existing at the point of `when`.

**Comment by daurnimator** [comment]
Don't revert in a PR. Just remove the commit.

**Comment by RyanSquared** [comment]
I figure leave the commits in the PR if someone else makes the comment, then rebase and drop before merging. But these are enough comments that no context is lost.

**Review by daurnimator — DISMISSED** [review]

**Comment by RyanSquared** [comment]
I'd like another review, and for someone else to be the one to merge _because I don't actually have write access_.

**Comment by RyanSquared** [comment]
@daurnimator feel free to merge whenever if no one else wants to review it.

**Comment by KellerFuchs** [comment]
@RyanSquared I rebased this, since it was conflicting with other stuff that got merged (#32 I guess).
I started reviewing a few days ago, but didn't have spare brains much; will try to finish by next weekend.

File: schema.sql (hunk 2 above)
Comment (KellerFuchs) [inline schema.sql]: `reserved_usernames`?  `banned_usernames` sounds like is it for users who misbehaved.

File: schema.sql [hunk 3]
```diff
@@ -62,6 +75,49 @@ create constraint trigger max_users
...
+create function check_banned_username() returns trigger
+    language plpgsql as $$
+    begin
+        if (exists(select 1 from banned_usernames where name = new.name)) then
+            raise check_violation using message = 'username banned: '||new.name;
+        end if;
+        return new;
+    end $$;
+create constraint trigger check_banned_username
+    after insert on passwd
```
Comment (KellerFuchs) [inline schema.sql:121]: You also need to check after an insert in `banned_usernames`

**Review by KellerFuchs — CHANGES_REQUESTED** [review]
Great work, thanks!

I left some actionable feedback already, and I suspect there might be refactor opportunities, but it's 2.40am and I need to snooze.

File: schema.sql (hunk 2 above)
Comment (KellerFuchs) [reply schema.sql]: Also, should the table get populated?

**Comment by KellerFuchs** [comment]
I forgot to mention in the review message, but all this needs tests.

**Comment by KellerFuchs** [comment]
BTW, I would have split that in individual commits, so it's easier to find which bit closed #27 for instance.

File: schema.sql (hunk 2 above)
Comment (RyanSquared) [reply schema.sql]: yeah, reserved makes more sense. Also, I seemed to have forgotten to push it, but I had a thing in the Makefile to insert a bunch of names from the list on the issue relating to this section. Will do later.

**Comment by RyanSquared** [comment]
Yeah, sorry, this was a bit weird of a PR. I basically wrote the code in the issue, realized afterwards "hey it actually works why not shove it all in a PR".

File: schema.sql (hunk 3 above)
Comment (RyanSquared) [reply schema.sql:121]: Why? It already checks with the `UNIQUE` constraint whether or not it already exists.

Comment (KellerFuchs) [reply schema.sql:121]: No, I meant that when inserting in `banned_usernames`, we should check that the username doesn't already exist.

Comment (RyanSquared) [reply schema.sql:121]: Oh, alright, will fix tomorrow hopefully?

File: schema.sql (hunk 2 above)
Comment (KellerFuchs) [reply schema.sql]: I renamed the table while splititng stuff into commits.

File: schema.sql (hunk 3 above)
Comment (KellerFuchs) [reply schema.sql:121]: I just fixed it

**Review by KellerFuchs — DISMISSED** [review]
OK, there is still stuff that needs doing after I took another pass at it:
- [ ] actually populate the `reserved_usernames` table;
- [ ] add tests (positive and negative) for all those constraints.

**Comment by KellerFuchs** [comment]
@daurnimator Can you have a look through my changes?

File: schema.sql [hunk 4]
```diff
@@ -62,6 +77,64 @@ create constraint trigger max_users
...
+        end if;
+	return new;
+    end $$;
+create constraint trigger check_name_exists_group
+    after insert or update on "group"
+    for each row
+    execute procedure check_invalid_name_for_group();
+
+-- prevent users from taking names which are typically reserved
+CREATE TABLE "reserved_usernames" (
```
Comment (daurnimator) [inline schema.sql]: case.

File: schema.sql [hunk 5]
```diff
@@ -62,6 +77,64 @@ create constraint trigger max_users
...
+    "name" username_t UNIQUE NOT NULL
+);
+create function check_reserved_username() returns trigger
+    language plpgsql as $$
+    begin
+        if (exists(select 1 from reserved_usernames where name = new.name)) then
+            raise check_violation using message = 'username banned: '||new.name;
```
Comment (daurnimator) [inline schema.sql]: s/banned/reserved/?

**Review by daurnimator — DISMISSED** [review]
Looks okay to me other than minor comments left.

**Comment by RyanSquared** [comment]
We can drop in a file of reserved usernames and find a way to populate it using the Makefile, right? Should be pretty simple.

**Comment by KellerFuchs** [comment]
@RyanSquared Yes, populating a table from a textfile is simple.

File: usernames.sql [hunk 6]
```diff
@@ -0,0 +1 @@
+copy reserved_usernames (name) from '/tmp/usernames.txt';
```
Comment (daurnimator) [inline usernames.sql]: Why is it in `/tmp`?
Shouldn't it be in e.g. current dir?

Comment (RyanSquared) [reply usernames.sql]: No, this for some reason isn't readable by postgresql. Plus, postgresql requires an absolute path, and copying to /tmp is cleaner.

File: Makefile [hunk 7]
```diff
@@ -22,6 +23,8 @@ install: $(SCHEMA_FILES)
 	createdb userdb
 	$(foreach file,$(SCHEMA_FILES),psql -v ON_ERROR_STOP=1 -h localhost -d userdb -f $(file);)
 
+reserve: usernames.txt usernames.sql
+	psql -v ON_ERROR_STOP=1 -h localhost -d userdb -f usernames.sql
```
Comment (daurnimator) [inline Makefile]: Could you add usernames.sql to SCHEMA_FILES so that it's done by `install` like everything else?

**Comment by KellerFuchs** [comment]
I merged `master` in to make CI run and solve the merge conflict.
This still needs tests.

**Comment by RyanSquared** [comment]
> This still needs tests.

And to be re-PR'd to avoid the disaster of a merge/rebase that happened before

File: schema.sql [hunk 8]
```diff
@@ -46,9 +46,25 @@ create table "aux_groups" (
...
 );
 
--- prevent creation/update of a user if the number of users
+-- prevent creation/update of a user/host if the number of users
 -- in the group 'users' that have that host
 -- is equal to the maxUsers for that host
+create function check_hosts_for_hosts() returns trigger
+    language plpgsql as $$
+    begin
+        if ((select count(*) from passwd where passwd.host = new.name) <= new.maxusers) then
```
Comment (VictorKoenders) [inline schema.sql]: Should this not be `>= new.maxusers` ?

**Comment by KellerFuchs** [comment]
@RyanSquared Not sure what you mean by “re PR'd”

File: schema.sql (hunk 8 above)
Comment (KellerFuchs) [reply schema.sql]: Yeap.
@RyanSquared That PR really needs to include tests.  I will push some for `reserved_usernames` but please do the rest.

**Review by KellerFuchs — CHANGES_REQUESTED** [review]
- [x] I added tests for `reserved_usernames`.
- [ ] Everything else needs tests.
- [ ] `check_hosts_for_hosts` needs fixing; also, it is a silly function name.

**Comment by KellerFuchs** [comment]
That's exciting.  :(
Here the check works like a charm.

**Comment by KellerFuchs** [comment]
@RyanSquared I did some git surgery on this feature branch, removed commits unintentionally picked from `master` and split your giant initial commit into 3 distinct ones, for the different enforced checks.

**Comment by KellerFuchs** [comment]
@RyanSquared Mind checking why the test fails in CI?

**Comment by RyanSquared** [comment]
- https://travis-ci.org/hashbang/userdb/builds/382269345#L566 (psql:reserved.sql:15: ERROR:  syntax error at or near "on")
- https://travis-ci.org/hashbang/userdb/builds/382269345#L728 (1. unit_tests.reserved_usernames() --> Successfully inserted invalid user..)

I'll look into the exact details (and how to fix the `on conflict do nothing`) in a bit.

**Comment by KellerFuchs** [comment]
@RyanSquared Reminder: you still need to write tests.

File: schema.sql (hunk 1 above)
Comment (RyanSquared) [reply schema.sql:106]: Only for update and I don't think we would ever update enough to be relevant

**Comment by RyanSquared** [comment]
@hashbang/administrators ping

**Comment by benharri** [comment]
No objections here.

**Review by daurnimator — APPROVED** [review]
I'm happy to merge now; we can always do renames and add more tests later.
This has been blocked for too long.
